### ISSUE TITLE
Implications of transaction support on COPY, DDL, etc.

### ISSUE THREAD

**Issue by jasonmp85** [citusdata/citus#666]
In #627, we added some tests to verify the expected behavior of `COPY` during transaction blocks (i.e. those delimited by `BEGIN` and `COMMIT`). In particular, `COPY` is presently permitted so long as no modifications have yet taken place within the block. If a block already has an `INSERT`, `UPDATE`, or `DELETE` command within it, the participants of the transactions are "locked" and no further connections are permitted during the block's duration (and `COPY` always opens new connections).

@marcocitus questioned this approach:

> There could be some problems with `COPY` in transactions. For example, you could get deadlocks on the advisory locks, since two transactions could lock shards in opposite order. For example, transaction A does `COPY` and locks shards 1-2-3, transaction B does `UPDATE` on shard 3 and then a COPY and locks shard 1-2-3. They could get stuck waiting for each other, though I guess PostgreSQL would detect a deadlock in this case.
> 
> You could also get a distributed deadlock on the row-level locks, e.g. if you're writing the same primary key in multiple transactions, though this is a more general and much harder to solve problem.
> 
> In any case, `COPY` does not roll back after the statement finishes, so maybe we should just disallow it in transaction blocks.

I agreed that it might be best to just disallow `COPY` in transaction blocks:

> That's what the latest DDL change did for DDLs. Adding a simple check to `COPY` could do the same and avoid all this reasoning about what is or is not allowed.

@marcocitus agreed and added some implementation details and context:

> Yeh, we added the check in DDL partly because of potential conflicts with single-shard transactions. I think adding the `PreventTransactionChain` thing to the `COPY` code path would make sense as well (e.g. in multi_utility.c, since the `isTopLevel` variable is available there).

I was ready to make this change, but @anarazel pointed out that the same reason we added `BEGIN`/`COMMIT` support at all (client libraries that default to emitting such blocks) is a reason we shouldn't disallow such blocks for `COPY` (or even DDL statements?)…

I kind of feel the deadlock concern is a bit of a red herring (clients should timeout and retry at some point anyways, and triggering such a deadlock would be rare), but was indifferent one way or the other.

So it appears there's some room for discussion around this point, which we can continue here.

**Comment by jasonmp85** [comment]
😈 THE ISSUE NUMBER 🔥 🙈

**Comment by sumedhpathak** [comment]
@jasonmp85 heh. You want this issue assigned to you? :)

**Comment by lfittl** [comment]
@jasonmp85 Just as a minor "what does the ORM user see" note since you mention DDL commands:

With current master (i.e. #627 merged) one now actually has to disable transactions for DDL changes, e.g. inside Rails migrations.

This is a change from 5.1 where that was allowed (even though not correctly behaving due to 2PC not being implemented).

**Comment by anarazel** [comment]
I don't think DDL and bulk load really are that comparable. The former will often not happen from the normal application, whereas the latter often will.

**Comment by jasonmp85** [comment]
> The former will often not happen from the normal application, whereas the latter often will.

Depends on the application. I suppose in the Citus context, bulk loads are of course more common. But I've _never_ used bulk load features of database drivers, yet executed hundreds of Rails migrations during deployments using our driver.

(I see @ifittl mentions the same, though IIRC, this change wasn't actually from #627, but from another change that added a transaction block check to our DDL logic).

**Comment by lfittl** [comment]
@jasonmp85 Might have been another recent change - apologies if I misattributed.

Do you think we can make transactions on the master + 2PC work for 5.2?

Also, should I open a separate issue for this?

**Comment by anarazel** [comment]
On 2016-07-22 13:43:48 -0700, Jason Petersen wrote:
> > The former will often not happen from the normal application, whereas the latter often will.
> 
> Depends on the application. I suppose in the Citus context, bulk loads are of course more common. But I've _never_ used bulk load features of database drivers, yet executed hundreds of Rails migrations during deployments using our driver.

Sure, but we're talking about citus here ;). But even so, it's not
unlikely to be more realistic to locally fix this for migrations, but
continue transactions for copy.

**Comment by jasonmp85** [comment]
> Do you think, for DDL changes, we can make transactions on the master + 2PC work for 5.2?
>
> Also, should I open a separate issue for this?

I don't understand what this is asking, so probably?

**Comment by marcocitus** [comment]
What if we introduce some mutually exclusive transaction modes? E.g. if you do a DML statement in a transaction, you can no longer do DDL and vice-versa, perhaps COPY and master_modify_multiple_shards should be separate modes.

**Comment by jasonmp85** [comment]
> What if we introduce some mutually exclusive transaction modes?

Planning on doing precisely this, though we'll actually have some sort of bit field with flags to designate whether the present transaction has yet performed a DDL, DML, etc. command. Logic will enforce the mutual exclusion property.

**Comment by marcocitus** [comment]
This seems no longer relevant, especially after #1455, which enabled DML+DDL+COPY in the same transaction.
